Updated Get Started with Messaging + (#401)
wormhole-foundation/wormhole-docs: tools/solidity-sdk/get-started.md

Reviewer: So the Solidity SDK is definitely a bit different from the TypeScript SDK, haha obviously, but we still need to treat it as tool documentation. So if I put myself in the shoes of someone new to the SDK and trying to get started with it, I'm going to want to know how to install the SDK, how to import the different contracts, what contracts are available to me to import (high-level), and then an example

Reviewer: Ok, so iof you put yourself in the shoes of someone who wants to build a contract-based solution and you just came across the Solidity SDK with no real context about what all it provides, what would you expect to find on the get started page?

For me, I think it'd be nice to see what's available on a high-level. If not on this page, the other option is the reference page since the tools don't have an overview page. Regardless, I think we should most definitely add a high-level list or link to one because this would be the type of question I would hope to find an answer to

Reviewer: Maybe in the introduction we need to elaborate a little bit more in this sentence: "providing all necessary Solidity interfaces along with useful libraries and tools for testing."

That's pretty vague and yes, it's an introduction, but what does this mean?

---
title: Get Started with the Solidity SDK
description: Follow this guide to deploy Wormhole Solidity SDK-based sender and receiver smart contracts and use them to send testnet USDC across chains.
categories: Basics, Solidity-SDK
---

# Get Started with the Solidity SDK

The [Wormhole Solidity SDK](https://github.com/wormhole-foundation/wormhole-solidity-sdk){target=\_blank} provides Solidity interfaces, prebuilt contracts, and testing tools to help Solidity developers build on-chain Wormhole integrations via smart contracts. You can use the [Wormhole TypeScript SDK](/docs/tools/typescript-sdk/get-started/){target=\_blank} for off-chain integrations without writing Solidity.

> ## Install the Wormhole Solidity SDK

Reviewer: To align with the TypeScript SDK page

---
title: Get Started with the Solidity SDK
description: Follow this guide to use the Wormhole Solidity SDK's interfaces and tools to help you quickly build on-chain integrations using smart contracts.
categories: Basics, Solidity-SDK
---

# Get Started with the Solidity SDK

The [Wormhole Solidity SDK](https://github.com/wormhole-foundation/wormhole-solidity-sdk){target=\_blank} provides Solidity interfaces, prebuilt contracts, and testing tools to help Solidity developers build on-chain Wormhole integrations via smart contracts. You can use the [Wormhole TypeScript SDK](/docs/tools/typescript-sdk/get-started/){target=\_blank} for off-chain integrations without writing Solidity.

## Install the Wormhole Solidity SDK

Use Foundry's [`forge`](https://book.getfoundry.sh/forge/){target=\_blank} to install the SDK using the following command:

```bash
forge install wormhole-foundation/wormhole-solidity-sdk
```

> 
> Next, review the following sections for more information on SDK components and features, or jump to [How It Works](#how-it-works) for a build-along demo project.
> 
> ## Key Components and Features
> 

Reviewer: ```suggestion

## Key Components

```

Then the rest of this section can be like the example I shared in the main comment on the PR. Very brief, cut out any extra stuff. Just quick like hey here's what's available type deal

The following key components and features work together to make your on-chain Wormhole integration easier to build.

> ### Wormhole Relayer Integration

Reviewer: Delete this section

- **Reduce operational burden**: No need to run relay infrastructure or manage gas on multiple chains.
- **Accelerate development with prebuilt contracts**: Common use cases like sending messages, tokens, and USDC come together quickly leveraging base contracts.

    ??? example "Solidity SDK base contracts"
        - [**`Base.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/Base.sol){target=\_blank}: Uses Wormhole interfaces to authorize and verify a registered sender.
        - [**`TokenBase.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/TokenBase.sol){target=\_blank}: Uses `TokenReceiver` and `TokenSender` contracts to define functions for transferring tokens.
        - [**`CCTPBase.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/CCTPBase.sol){target=\_blank}: Uses `CCTPSender` and `CCTPReceiver` contracts to define functions for transferring USDC.

> ### Solidity Interfaces

Reviewer: Delete

??? example "Solidity SDK interfaces"
    - [**`ITokenBridge.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/interfaces/ITokenBridge.sol){target=\_blank}: Defines key structs and functions for token attestation, wrapping and transferring tokens, monitoring transaction progress.
    - [**CCTP Interfaces**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/tree/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/interfaces/CCTPInterfaces){target=\_blank}: A set of interfaces for USDC transfers via CCTP for sending, relaying, and receiving messages and tokens.
    - [**`IWormholeReceiver.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/interfaces/IWormholeReceiver.sol){target=\_blank}: Defines the `receiveWormholeMessages` function.
    - [**`IWormholeRelayer.sol`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/interfaces/IWormholeRelayer.sol){target=\_blank}: Defines key structs and functions to identify, send, and deliver messages and follow the progress of transactions.

> ### Helpful Constants

Reviewer: Delete

Auto-generated Solidity constants help avoid manual entry errors and ensure consistent delivery.

??? example "Solidity SDK constants"
    - [**Wormhole Chain ID's**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/Chains.sol){target=\_blank}: Generated list of Wormhole Chain ID's for supported chains.
    - [**Circle CCTP Domain IDs**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/b9e129e65d34827d92fceeed8c87d3ecdfc801d0/src/CCTPAndTokenBase.sol){target=\_blank}: Generated list of defined CCTP domain ID's to ensure USDC transfers use the correct domain for a given chain. 
    - [**`chainConsts.ts`**](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/75ddcec06ffe9d62603d023357caa576c5ea101c/gen/chainConsts.ts){target=\_blank}: Returns values to identify properties and contract addresses for each supported chain.

> ### Testing Utilities

Reviewer: Delete

The Wormhole Solidity SDK also includes a robust set of [testing utilities](https://github.com/wormhole-foundation/wormhole-solidity-sdk/tree/75ddcec06ffe9d62603d023357caa576c5ea101c/test) for simulating message and token transfers to test your build.

> ## Prerequisites
> 
> The following demo illustrates the use of Wormhole Solidity SDK-based smart contracts to send testnet USDC between supported chains.
> 

Reviewer: ```suggestion
## Example Usage

The following demo illustrates the use of Wormhole Solidity SDK-based smart contracts to send testnet USDC between supported chains.

### Prerequisites
```

haha change my mind on this, sorry

Before you begin, ensure you have the following:

- [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm){target=\_blank} installed
- [TypeScript](https://www.typescriptlang.org/download/){target=\_blank} installed
- [Foundry](https://book.getfoundry.sh/getting-started/installation){target=\_blank} installed

> - Testnet tokens for two supported chains. This example uses [testnet AVAX for Avalanche Fuji](https://core.app/tools/testnet-faucet/?subnet=c&token=c){target=\_blank} and [testnet CELO for Celo Alfajores](https://faucet.celo.org/alfajores){target=\_blank} and can be adapted to any supported chains.

Reviewer: ```suggestion
- Testnet tokens for two supported chains. This example uses [testnet AVAX for Avalanche Fuji](https://core.app/tools/testnet-faucet/?subnet=c&token=c){target=\_blank} and [testnet CELO for Celo Alfajores](https://faucet.celo.org/alfajores){target=\_blank} and can be adapted to any supported chains
```

- [USDC testnet tokens](https://faucet.circle.com/){target=\_blank} on your source chain for cross-chain transfer

> 

Reviewer: ```suggestion

### Set Up a Project

```

Follow these steps to prepare your development environment:

1. Create a directory for your project, navigate into it, and install the Wormhole Solidity SDK: 

    ```bash
    mkdir solidity-token-transfer
    cd solidity-token-transfer
    forge install wormhole-foundation/wormhole-solidity-sdk
    ```

2. Install dependencies for use with your transfer script, including the Wormhole TypeScript SDK, and initiate a new Node.js project:

    ```bash
    npm init -y && npm install @wormhole-foundation/sdk ethers -D tsx typescript
    ```

> ## Create and Deploy Contracts

Reviewer: ```suggestion
### Create and Deploy Contracts
```

This project uses sender and receiver contracts to access the `WormholeRelayer` interface's [`TokenSender`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/baa085006586a43c42858d355e3ffb743b80d7a4/src/WormholeRelayer/TokenBase.sol#L24){target=\_blank} and [`TokenReceiver`](https://github.com/wormhole-foundation/wormhole-solidity-sdk/blob/baa085006586a43c42858d355e3ffb743b80d7a4/src/WormholeRelayer/TokenBase.sol#L147){target=\_blank} base classes to simplify sending tokens across chains.

Follow these steps to create and deploy your sender and receiver Solidity contracts:

1. Use the following example code to create `CrossChainSender.sol`:

    ```solidity title="CrossChainSender.sol"
    --8<-- "code/tools/solidity-sdk/get-started/solidity-sdk-1.sol"
    ```

>     This contract assigns `CrossChainSender` to the `TokenSender` role, initializes the contract, calculates a cost of transfer estimate, defines transfer parameters, and initiates the transfer using the `sendTokenWithPayloadToEvm` function from `WormholeRelayer`.

Reviewer: ```suggestion
    This contract extends `TokenSender`, gaining access to its functionality. It initializes the contract with the required addresses, calculates estimated transfer costs, defines transfer parameters, and initiates the transfer using the `sendTokenWithPayloadToEvm` function from `WormholeRelayer`.
```

Instead of "assign" and "role"

2. Use the following example code to create `CrossChainReceiver.sol`:

    ```solidity title="CrossChainSender.sol"
    --8<-- "code/tools/solidity-sdk/get-started/solidity-sdk-2.sol"
    ```

>     This contract assigns `CrossChainReceiver` to the `TokenReceiver` role, initializes the contract, receives the payload and tokens, verifies the transfer is from a registered sender, decodes the recipient address, and transfers the tokens to the recipient.

Reviewer: ```suggestion
    This contract extends `TokenReceiver`, gaining access to its functionality. It initializes the contract with the required addresses, receives the payload and tokens, verifies the transfer is from a registered sender, decodes the recipient address, and transfers the tokens to the recipient.
```

Same thing here

3. Deploy the contracts using your preferred deployment method. Make sure you deploy `CrossChainSender.sol` to your desired source chain and `CrossChainReceiver.sol` to the target chain. Save the deployed contract addresses for each contract. You will need them for your transfer script.

> ## Create Your Transfer Script

Reviewer: ```suggestion
##  Use Contracts to Transfer USDC
```

What are we doing with the script though? What's the important part here

1. Once your contracts are deployed, create a `transfer.ts` file to handle the multichain transfer logic:

    ```bash
    touch script/transfer.ts
    ```

2. Set up secure access to your wallets. This guide assumes you are loading your private key(s) from a secure keystore of your choice, such as a secrets manager or a CLI-based tool like [`cast wallet`](https://book.getfoundry.sh/reference/cast/cast-wallet){target=\_blank}.

    !!! warning
        If you use a `.env` file during development, add it to your `.gitignore` to exclude it from version control. Never commit private keys or mnemonics to your repository.

3. Open `transfer.ts` and add the following code:

    ```typescript title="transfer.ts"
    --8<-- "code/tools/solidity-sdk/get-started/solidity-sdk-3.ts"
    ```

>     This script defines sender and receiver contract addresses, fetches needed ABI information, creates a connected signer, converts decimals, qoutes transfer cost, and initiates the token transfer.

Reviewer: ```suggestion
    This script defines the sender and receiver contract addresses, fetches the necessary ABI information, creates a connected signer, converts decimals, calculates the estimated transfer cost, and initiates the token transfer.
```

---
title: Get Started with the Solidity SDK
description: Follow this guide to deploy Wormhole Solidity SDK-based message sender and receiver smart contracts and use them to send messages across chains.
categories: Basics, Solidity-SDK
---

# Get Started with the Solidity SDK

:simple-github: [Source code on GitHub](https://github.com/wormhole-foundation/demo-wormhole-messaging){target=\_blank}

## Introduction

Wormhole's messaging protocol simplifies sending data, triggering events, and initiating transactions across blockchain networks. This guide demonstrates how to configure and deploy contracts to send messages from Avalanche Fuji to Celo Alfajores.

## Prerequisites

Before you begin, make sure you have the following:

- [Node.js and npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm){target=\_blank}
- [Foundry](https://book.getfoundry.sh/getting-started/installation){target=\_blank} for deploying contracts and encrypting your private key
- [Testnet AVAX for Avalanche Fuji](https://core.app/tools/testnet-faucet/?subnet=c&token=c){target=\_blank}
- [Testnet CELO for Celo Alfajores](https://faucet.celo.org/alfajores){target=\_blank}
- Wallet private key

> ## Install and Set Up Project

Reviewer: What about for people that don't want to set up a project, maybe they already have one. They're just interested in getting started with the SDK specifically. We're not including any instructions for them to install the SDK

## Send Your First Message

Follow these steps to use your deployed contracts and send your first message:

1. Run the `sendMessage.ts` script using the following command:

    ```bash
    npm run send:message
    ```

2. Enter your Foundry keystore password in the terminal when prompted

3. You will see terminal output similar to the following:

    --8<-- "code/tools/solidity-sdk/get-started/terminal-output-05.html"

4. You can also use the [Celo Alfajores Testnet Explorer](https://alfajores.celoscan.io/){target=\_blank} to view your `MessageReceiver` contract. Select **Events** to see the events emitted by the contract when your message was received. You can use the dropdowns to change **Hex** to **Text** to read the message. It will look similar to the image below:

    ![Contract events on Celo Alfajores Testnet Explorer](/docs/images/tools/solidity-sdk/get-started/messaging-contracts01.webp)

Congratulations! You've successfully sent and received a message across networks using Wormhole Solidity SDK-based smart contracts. 

## Next Steps

> - [Fetch the Signed VAA](TODO WIP){target=\_blank} - whether your message is sent using core protocol or custom smart contracts, follow this guide to use your transaction information to fetch a signed VAA and decode the payload

Reviewer: ```suggestion
- [**Fetch the Signed VAA**](TODO WIP){target=\_blank} - whether your message is sent using core protocol or custom smart contracts, follow this guide to use your transaction information to fetch a signed VAA and decode the payload
```
